### PR TITLE
Student blog: GSoC results

### PR DESCRIPTION
**Pull request by vasalf**

### CODE REVIEW

File: content/news/2019-08-26-gsoc-results.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
+---
+title: 'Student Blog: GSoC Results'
+author: 'Vasily Alferov'
+---
+
+I am a student at Summer of Haskell this year, and that's my final post. This
+post is written to sum up the work I've done. Although this post is a formal
+requirement of GSoC, I hope that this post is attractive to everyone.
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
requirement of GSoC, I wanted to publish it here to reach a wider audience.
```

File: content/news/2019-08-26-gsoc-results.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+# My proposal
+
+When you apply to Summer of Code, you write a proposal. The proposal is a
+document in which you describe your ideas on the chosen project. It should be a
+clear, detailed text with suggestions on every subtask. The proposal should also
+include a timeline, in which you estimate the time you spend on every subtask.
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
include a timeline, in which you estimate the time you intend to spend on each of those subtasks.
```

File: content/news/2019-08-26-gsoc-results.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+I chose [this](https://summer.haskell.org/ideas.html#algebraic-graphs) project
+for my summer. In my proposal, I drafted all the algorithms mentioned in the
+list and suggested a few more. I published this part of my proposal as a Github
+gist [there](https://gist.github.com/vasalf/c246171d8ebd784487624853566b8df2).
+
+I don't suggest this gist as an example of a good proposal. It's only a part of
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
I don't suggest this gist as a complete example of a good proposal: it's only a part of
```

File: content/news/2019-08-26-gsoc-results.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+the document I submitted. You should also include some information about you,
+together with the timeline. Communication with your future mentors is also a
+significant part of the application.
+
+However, as I mentioned in
+[one of my previous posts](https://summer.haskell.org/news/2019-05-29-alga-introduction.html),
+another student is doing the part suggested in the ideas list. My task is to
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
another student ended up doing the part suggested in the ideas list. So my task is to
```

File: content/news/2019-08-26-gsoc-results.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+This task was my idea. I mentioned it in my proposal. I meant that finding
+maximum matchings in bipartite graphs should be easily implemented when we have
+algorithms for finding maximum flows in networks. Kuhn's algorithm is an
+application of the Ford-Fulkerson algorithm, and the Hopcroft-Karp algorithm is
+an application of Dinic's algorithm.
+
+However, this option is not the best. Both algorithms have specialized
+implementations that work times faster. So my task for this summer was to
+introduce bipartite graphs and special functions for them.
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
introduce bipartite graphs and special functions for working with them.
```

File: content/news/2019-08-26-gsoc-results.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+`test/` directory. The documentation for each function and datatype precedes the
+declaration. After release, it will compile to beautiful Haddock file like
+[this](http://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph.html).
+
+# Part I. Definition and properties.
+
+**Link to PR:** [https://github.com/snowleopard/alga/pull/207](https://github.com/snowleopard/alga/pull/207)
+
+In this part, I defined the `Bipartite.AdjacencyMap` datatype and added many
+functions for its properties.
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
functions to work with adjacency maps.
```

File: content/news/2019-08-26-gsoc-results.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+The datatype represents a map of vertices into their neighbours. I defined it a
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
The datatype represents a map of vertices into their neighbours. I defined it as
```

File: content/news/2019-08-26-gsoc-results.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+two maps:
+
+```Haskell
+data AdjacencyMap a b = BAM {
+    leftAdjacencyMap :: Map.Map a (Set.Set b),
+    rightAdjacencyMap :: Map.Map b (Set.Set a)
+}
+```
+
+The properties are based on the
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
The properties are based on the existing
```

File: content/news/2019-08-26-gsoc-results.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+[properties](http://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph.html)
+of graphs in Alga.
+
+# Part II: Testing bipartiteness
+
+**Link to PR:** [https://github.com/snowleopard/alga/pull/218](https://github.com/snowleopard/alga/pull/218)
+
+There is a folklore algorithm that tests a given graph on being bipartite. The
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
There is a folklore algorithm that checks if a given graph is bipartite. The
```
active tense reads a little easier than passive

File: content/news/2019-08-26-gsoc-results.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+task to implement this algorithm in Haskell was a little challenge for me.
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
task to implement this algorithm in Haskell was a little challenging for me.
```

File: content/news/2019-08-26-gsoc-results.md [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+I finished up with the following definition of the function:
+
+```Haskell
+detectParts :: Ord a => AM.AdjacencyMap a -> Either (OddCycle a) (AdjacencyMap a a)
+```
+
+This naturally means that the graph is either bipartite or contains an odd
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
As you can clearly see in the type, this means that the graph is either bipartite or contains an odd
```

File: content/news/2019-08-26-gsoc-results.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+cycle.
+
+The implementation is so exciting that I wrote a whole
+[post](https://summer.haskell.org/news/2019-07-26-testing-bipartiteness.html)
+about it. I explained the reason I needed monad transformers there and made some
+interesting benchmarks that pointed me to use the explicit `INLINE` directive.
+
+# [WIP] Part III. Graph families.
+
+**Link to the unfinished PR**: [https://github.com/snowleopard/alga/pull/226](https://github.com/snowleopard/alga/pull/226)
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
**Link to the unfinished PR**: <https://github.com/snowleopard/alga/pull/226>
```

File: content/news/2019-08-26-gsoc-results.md [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+Each PR contains the actual implementation, tests, and documentation. The whole
+project is release-ready after merging each one of them. I put the tests in the
+`test/` directory. The documentation for each function and datatype precedes the
+declaration. After release, it will compile to beautiful Haddock file like
+[this](http://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph.html).
+
+# Part I. Definition and properties.
+
+**Link to PR:** [https://github.com/snowleopard/alga/pull/207](https://github.com/snowleopard/alga/pull/207)
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
**Link to PR:** <https://github.com/snowleopard/alga/pull/207>
```

File: content/news/2019-08-26-gsoc-results.md [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+}
+```
+
+The properties are based on the
+[properties](http://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph.html)
+of graphs in Alga.
+
+# Part II: Testing bipartiteness
+
+**Link to PR:** [https://github.com/snowleopard/alga/pull/218](https://github.com/snowleopard/alga/pull/218)
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
**Link to PR:** <https://github.com/snowleopard/alga/pull/218>
```

File: content/news/2019-08-26-gsoc-results.md [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+```Haskell
+path :: (Ord a, Ord b) => List a b -> AdjacencyMap a b
+```
+
+As for now, the PR is almost merge-ready, only several small notices need fixes.
+
+# [WIP] Part IV. Hopcroft-Karp algorithm for maximum matchings
+
+**Link to the unfinished PR**: [https://github.com/snowleopard/alga/pull/229](https://github.com/snowleopard/alga/pull/229)
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
**Link to the unfinished PR**: <https://github.com/snowleopard/alga/pull/229>
```

File: content/news/2019-08-26-gsoc-results.md [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+data List a b = Nil | Cons a (List b a)
+```
+
+So the `path` definition is:
+
+```Haskell
+path :: (Ord a, Ord b) => List a b -> AdjacencyMap a b
+```
+
+As for now, the PR is almost merge-ready, only several small notices need fixes.
```
Comment (jaspervdj) [inline content/news/2019-08-26-gsoc-results.md]:
```suggestion
As for now, the PR is almost merge-ready, only several small comments need fixes.
```

**Review by jaspervdj — APPROVED** [review]
Looks good, I added a number of small suggestions

File: content/news/2019-08-26-gsoc-results.md (hunk 11 above)
Comment (vasalf) [reply content/news/2019-08-26-gsoc-results.md]: I changed this paragraph to
```
It is known that a graph is bipartite if and only if it contains no cycles of
odd length. This function either finds an odd cycle or returns a partition.
```

**Pull request merged**